Exclude user generated tags in search query

alexef opened the pull request:
Basically, allow tag names in the query parser to be prefixed with "-" (for instance, "-level:error"). Then teach the search backend to split the tags into two categories: filtered and excluded, and modify the subquery accordingly. Also update autocomplete so that "-is:" or "-has:" are not accepted.

No unit tests (yet).

Fixes: #2646

dcramer commented:
I think in general the direction looks good. We'd definitely need tests to support it, but I'll leave it to others to actually review the changes.

alexef commented:
@tkaemming @dcramer Kind of a late update, but now the PR also brings a basic unit test. Any chance to get this merged?

dcramer commented:
We might consider using ! Instead of - which would probably read more logically if we end up applying it to other things. No strong opinion though and I'm going to defer to @tkaemming and @mattrobenolt on review.

tdtm commented:
The minus is consistent with existing search engines (github, Mysql full
text, Google,...) rather than usual programming languages.

What "other things" do you have in mind?

Le mer. 22 févr. 2017 20:24, David Cramer <notifications@github.com> a
écrit :

We might consider using ! Instead of - which would probably read more
logically if we end up applying it to other things. No strong opinion
though and I'm going to defer to @tkaemming <https://github.com/tkaemming>
and @mattrobenolt <https://github.com/mattrobenolt> on review.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/getsentry/sentry/pull/4702#issuecomment-281773815>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AHZBfANdcp6QbUBFPr025FO6T6LGQqXLks5rfItagaJpZM4LSzp3>
.

dcramer commented:
i dont have a strong opinion on - or !

alexef commented:
Up :)

mattrobenolt commented:
Sorry about that, I'll make sure I give some feedback here this week.

tkaemming commented on src/sentry/search/utils.py:
```diff
@@ -236,9 +236,10 @@ def tokenize_query(query):
             continue
 
         key, value = token.split(':', 1)
+        real_key = key[1:] if key and key[0] == '-' else key
```
This can be moved into the conditional below since it's only used within that block.

tkaemming commented on src/sentry/search/django/backend.py:
```diff
@@ -136,12 +136,22 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
-                id__in=matches,
-            )
+            filter_tags = {k: v for k, v in six.iteritems(tags) if k[0] != '-'}
+            exclude_tags = {k[1:]: v for k, v in six.iteritems(tags) if k[0] == '-'}
```
This is a minor nitpick, but this double iteration of `tags` should probably just be a single loop that adds to `filter_tags` or `exclude_tags` based on the prefix. It'd probably make sense to also define a behavior if the tag is included as both an inclusion and exclusion filter. I guess it'd be excluded by default as it stands now?

alexef replied:
Adding a special case. I think a query with both filter and exclude for the same tag is broken, therefor returning queryset.none() makes sense here.

tkaemming commented on src/sentry/search/django/backend.py:
```diff
@@ -136,12 +136,22 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
+            if exclude_tags:
+                matches = self._tags_to_filter(project, exclude_tags)
+                if not matches:
+                    return queryset.none()
```
I don't think this logic makes sense here for a few reasons:

1. If there no matches because a tag doesn't exist, or a tag has no recent values, `_tags_to_filter` will return `None`. This mean that if there is nothing to exclude, the search will return no results.
2. `_tags_to_filter` assumes we're reducing the possible input set with each tag — for example, if the input set was `[1,2,3,4,5]` and the filters represented "all even numbers" and "numbers divisible by 4 with no remainder", the only item in the input set that matches all constraints is `4`. This isn't how exclusion filters typically work, though — if we were to *exclude* values using the exact same filter criteria, the expected result should be `[1, 3, 5]` since `2` is an even number and `4` is both an even number along with evenly divisible by `4`. The fact that `2` isn't evenly divisible by `4` isn't relevant. Instead of excluding values that contain *all* of the excluded tag values here, we'd actually want to exclude values that contain *any* of the included tag values. This way you can write filters like `-level:info -level:debug`. As it's written right now, a query of `-level:info -level:debug` would only exclude groups that have been tagged with *both* of those values.

tkaemming commented on src/sentry/search/django/backend.py:
```diff
@@ -136,12 +136,22 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
+                    id__in=matches,
+                )
+            if exclude_tags:
+                matches = self._tags_to_filter(project, exclude_tags)
```
`_tags_to_filter` will [return at most `1000` group primary keys](https://github.com/getsentry/sentry/blob/master/src/sentry/search/django/backend.py#L62), so this will possibly  result in false positives making it through this exclusion filter depending on how many groups exist within a project. I'm unsure if this is a realistic issue or not. My personal opinion is that we shouldn't provide a search that's intentionally going to return incorrect results, but we already don't necessarily return the correct results with the inclusion filters, so I'm not really sure how to proceed here.

alexef replied:
Any reason for that 1000 limit? Since it's the same behaviour, we can keep it broken, or remove the limit and fix it for both inclusion and exclusion.

tkaemming replied:
I think that it was a preventative measure to avoid a potentially expensive join when a tag is present on a very large number of distinct groups, but there's no rationale in the original change (getsentry/sentry#3486) so I suppose that's a @dcramer question.

alexef commented:
@tkaemming Looking deeper in the code, it seems like now it's not possible to pass more than one value for a tag - this gets overriden in parse_query, only the last one is kept: https://github.com/getsentry/sentry/blob/6ba6d960b0e5bcb5e0598f5f8c8f49cba96d3142/src/sentry/search/utils.py#L353

I'll dig some more.

alexef commented:
Just a note, my last commit addresses the comments, but it's breaking the tests. I'm currently reconfiguring the local setup, I find it painful, both with Docker or the local installation as in the docs.

Docker is incredibly slow for me when running the test suite, and with the local installation I'm still not able to finish setting up the environment, fixing issues one by one, last being `UnknownExtra: sentry 8.15.0.dev0 has no such extra feature 'dsym'`. I'm quite frustrated, I used to be happy with the sentry dev environment.

alexef commented on src/sentry/search/django/backend.py:
```diff
@@ -139,12 +139,33 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
+                else:
+                    filter_tags[k] = v
+            # Don't allow the same tag to be both excluded and included
+            common_tags = (k for k in filter_tags if k in exclude_tags)
```
You are right, I wasn't paying attention either. The convention makes sense, as long as we stick to the convention. Pushing a fix.

tkaemming commented:
f6c19eb045a296b2e67d78997616a4c5c5c12ac3 looks good. We'll still need to figure out about the [arbitrary limit bits](https://github.com/getsentry/sentry/pull/4702/files/679500df5b4b641a0967267aafde6343cca9bc67#r105790781) but otherwise this is looking promising.

alexef commented:
@tkaemming @dcramer rebased against master, but I think recent changes in master make `reserved_tag_names` in `sentry/search/utils.py` obsolete.

How can we proceed with this PR?

tkaemming commented on src/sentry/search/django/backend.py:
```diff
@@ -139,12 +139,37 @@ def _build_queryset(self, project, query=None, status=None, tags=None,
+                        continue
+                    all_matches += matches
+                queryset = queryset.exclude(
+                    id__in=set(all_matches),
```
- It might make sense to do this set difference in memory between inclusion and exclusion tags rather than deferring to the database.
- It may also make sense that if we are dealing with inclusion tags to provide those IDs to the `_tags_to_filter` method here so that we don't get back results for groups that we were not part of the result set in the first place.
